Fix/3557 scanner subfolder always scanning

Fesaa opened the pull request:
# Fixed
- Fixed: Scanner scanning subfolders of a series when no changes have been made (Fixes #3557)

Still work todo, PR is easier to track

Copied from discord:
If one subfolder changes, all are scanned. This is important otherwise the other subfolders content are removed. Should be a fine trade-off. 

This should be safe to do here, as if it matches a wrong series, we know this series is its parent and thus will always have changes. 

I have tested this on Windows & MacOS. Haven't had the chance to build for my server (Linux) yet. That said, I have no idea how the Comic stuff works etc. So... I've only tested with Manga as that's what I had laying around in the right layout to just copy over. 

TODO:

- [x] Unit Tests

Fesaa committed:
Potentially fix the subfolder issue

I have to test with image library & on windows and linux, as I think
modify time works differently there.

Fesaa committed:
Fix the fix on Windows

majora2007 commented on API/Services/Tasks/Scanner/ParseScannedFiles.cs:
```diff
@@ -222,6 +227,21 @@ private bool HasSeriesFolderNotChangedSinceLastScan(IDictionary<string, IList<Se
+            return seriesList;
+        }
+
+        return TryGetSeriesList(seriesPaths, _directoryService.GetParentDirectoryName(directory));
```
You need another short circuit if directory is the library root so we don't go out of bounds in some weird case.

Fesaa replied:
This case will be hit for every new series I'd image 😅

Fesaa committed:
Break loop when library root is reached

Fesaa committed:
Unit tests

Check that all series subfolders are marked as HasChanged

Maybe check Publisher layout, I don't use this. So I'm not sure that I did it correctly.

Fesaa committed:
Use comic library type & proper fix for relative paths

majora2007 commented on API.Tests/Services/ParseScannedFilesTests.cs:
```diff
@@ -349,4 +357,101 @@ public async Task ProcessFiles_ShouldCallFolderActionOnce()
+        var testDirectoryPath = Path.Join(
+            Path.Join(Directory.GetCurrentDirectory(), "../../../Services/Test Data/ScannerService/ScanTests"),
+            testcase.Replace(".json", string.Empty));
+        testDirectoryPath = Path.GetFullPath(testDirectoryPath);
```
Instead of applying this here, instead apply it on the return of ScannerHelper.GenerateTestDirectory() so that you can remove this library.Folders hack and all other tests will work for you.

majora2007 commented on API.Tests/Services/ParseScannedFilesTests.cs:
```diff
@@ -349,4 +357,101 @@ public async Task ProcessFiles_ShouldCallFolderActionOnce()
+
+        Assert.NotNull(postLib);
+        Assert.Equal(4, postLib.Series.Count);
+        var spiceAndWolf = postLib.Series.First(x => x.Name == "Spice and Wolf");
```
This block of code is hard to parse through. Break up each fetch/assert by a new line.

majora2007 commented on API.Tests/Services/ParseScannedFilesTests.cs:
```diff
@@ -349,4 +357,101 @@ public async Task ProcessFiles_ShouldCallFolderActionOnce()
+
+        Thread.Sleep(1100); // Ensure at least one second has passed since library scan
+
+        var executionerCopyDir =
```
I would add a comment to explain what this is doing and validating against. Something like:

// Add a new file to X directory and scan, validate that only Y and Z are showing as having had changes.

majora2007 commented on API.Tests/Services/ParseScannedFilesTests.cs:
```diff
@@ -349,4 +357,101 @@ public async Task ProcessFiles_ShouldCallFolderActionOnce()
+        var changes = res.Where(sc => sc.HasChanged).ToList();
+        Assert.Equal(2, changes.Count);
+        // Only volumes of The Executioner and Her Way of Life should be marked as HasChanged (Spice and Wolf also has 2 volumes dirs)
+        Assert.Equal(2, changes.Count(sc => sc.Folder.Contains("The Executioner and Her Way of Life")));
```
You need to run the proper scanner as well to validate that with your changes, the nothing is removed and that the changes are properly ingested. This can be in ScannerServiceTests with the same json file.

Fesaa replied:
I've done this for the test with sub folders. Does the publisher layout also need this test or not?

majora2007 replied:
No. Subfolders is the publisher layout.

majora2007 commented on API/Services/Tasks/Scanner/ParseScannedFiles.cs:
```diff
@@ -721,7 +752,8 @@ private async Task ParseFiles(ScanResult result, IDictionary<string, IList<Serie
         if (!result.HasChanged)
         {
-            result.ParserInfos = seriesPaths[normalizedFolder]
+            // We are certain TryGetSeriesList will return a valid result here, if the series wasn't present yet. It will have been changed.
```
I don't really understand this comment, can you clarify it a bit?

Fesaa replied:
Since I'm forcing the result of the new method to be not null, I've added the comment to explain this is safe to do. And will not throw an exception.

As all series that would return null, will have HasChanged set to true

Fesaa committed:
Clean up tests & Add full scan test
